src/date.c

```diff
@@ -1,18 +1,1012 @@
+                            fbreaktime breaktimefunc, fmaketime maketimefunc) {
+  /* 0 = year, 1 = month , 2 = date , 3 = hours, 
+                4 = minutes, 5 = seconds, 6 = ms */
+  struct timeparts tp;
+  unsigned long i;
+  
+  int* tp_arr[7]; /* C89 doesn't handle initialization 
+                          like x = {&tp.year, &tp.month, .... } */
+  tp_arr[0] = &tp.year; tp_arr[1] = &tp.month; tp_arr[2] = &tp.day;
+  tp_arr[3] = &tp.hour; tp_arr[4] = &tp.min; tp_arr[5] = &tp.sec;
+  tp_arr[6] = &tp.msec;
+  
+  memset(&tp, 0, sizeof(tp));
+  
+  if(breaktimefunc != NULL) {
+    breaktimefunc(current, &tp);
+  }
+  
+  for(i = 0; i < ARRAY_SIZE(tp_arr); i++) {
+    if(!d_istimeinvalid(&a->args[i])) {
+      *tp_arr[i] = (int)a->args[i];
+    }
+  }
+  
+  return maketimefunc(&tp);
+}
+
+static etime_t d_time_number_from_tp(struct v7 *v7, val_t this_obj, val_t args,
+            int start_pos, fbreaktime breaktimefunc, fmaketime makefilefunc) {
+  etime_t ret_time = INVALID_TIME;
+  long cargs;
+  
+  if((cargs=v7_array_length(v7, args)) >= 1
+                    && i_value_of(v7, this_obj) != V7_TAG_NAN) {
+    int i;
+    struct dtimepartsarr a = {{INVALID_TIME, INVALID_TIME, INVALID_TIME,
+                    INVALID_TIME, INVALID_TIME, INVALID_TIME, INVALID_TIME}};
+    etime_t new_part = INVALID_TIME;
+    
+    for(i=0; i < cargs && (i+start_pos < tpmax); i++) {
+      val_t argi = v7_array_at(v7, args, i);
+      if(!d_argtoint(v7, &argi, &new_part)) {
+        break;
+      }
+      
+      a.args[i+start_pos] = new_part;
+    }
+    
+    if(!d_istimeinvalid(&new_part)) {
+      etime_t current_time = v7_to_double(i_value_of(v7, this_obj));
+      ret_time = d_changepartoftime(&current_time, &a,
+                                          breaktimefunc, makefilefunc);
+    }
+  }
+  
+  return ret_time;
+}
+
+static val_t d_setTimePart(struct v7 *v7, val_t this_obj, val_t args,
+            int start_pos, fbreaktime breaktimefunc, fmaketime makefilefunc) {
+  val_t n;
+  etime_t ret_time = d_time_number_from_tp(v7, this_obj, args, start_pos,
+                                                breaktimefunc, makefilefunc);
+                                                                       
+  n = v7_create_number(ret_time);
+  v7_set_property(v7, this_obj, "", 0, V7_PROPERTY_HIDDEN, n);
+                                                                       
+  return n;
+}
+
+/*++++++++++++++ API +++++++++++++++*/
+
+/* constructor */
+static val_t Date_ctor(struct v7 *v7, val_t this_obj, val_t args) {
+  etime_t ret_time = 0;   
+  if(v7_is_object(this_obj) && this_obj != v7->global_object) {
+    long cargs = v7_array_length(v7, args);
+    if(cargs <=0 ) {
+      /* no parameters - return current date & time */
+      d_gettime(&ret_time);
+    } else if(cargs == 1) {
+      /* one parameter */
+      val_t arg = v7_array_at(v7, args, 0);
+      if(v7_is_string(arg)){ /* it could be string */
```

Comment:
What if it's a string but doesn't parse as a time. Shouldn't it be converted to int?